indexing-service/src/main/java/org/apache/druid/indexing/overlord/ZkWorker.java

```diff
@@ -63,6 +67,26 @@ public ZkWorker(Worker worker, PathChildrenCache statusCache, final ObjectMapper
     this.statusCache = statusCache;
     this.cacheConverter = (ChildData input) ->
         JacksonUtils.readValue(jsonMapper, input.getData(), TaskAnnouncement.class);
+    this.taskIdExtractor = createTaskIdExtractor(jsonMapper);
+  }
+
+  static java.util.function.Function<ChildData, String> createTaskIdExtractor(final ObjectMapper jsonMapper)
+  {
+    return (ChildData input) -> {
+      try {
+        JsonParser parser = jsonMapper.getFactory().createParser(input.getData());
+        while (parser.nextToken() != JsonToken.END_OBJECT) {
+          if ("id".equals(parser.getCurrentName())) {
+            parser.nextToken();
+            return parser.getText();
+          }
+        }
+        return null;
```

Comment:
what if a `null` is returned? will it result in exception in the `getRunningTaskIds`?

Reply:
It will not result in an exception in `getRunningTaskIds`, but it would allow `null` to be in the result set. I think from this class perspective is valid (if a task exists with a null id, then `getRunningTaskIds` should include null). And, `TaskAnnouncement` currently would (should?) exhibit this behaviour on deserialization.

However, I am not clear on the right behaviour here. If all tasks should have valid Ids, then I think the clearer least-surprising behaviour is to explicitly throw an exception, or maybe filter `null` elements out of the return value from `getRunningTaskIds`.

What do you think?

Reply:
I don't know if there's a case that there's no `id` field in the task. Maybe it only exists in our code here in theory. Since here is a performance critical path based on your investigation, I think we can leave the null here.